// @neo4j/graphql/dist/classes/subscription/cdc/cdc-api.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CDCApi = void 0;
const cypher_builder_1 = __importDefault(require("@neo4j/cypher-builder"));
const neo4j_driver_1 = require("neo4j-driver");
const error_has_gql_status_1 = require("../../../utils/error-has-gql-status");
class CDCApi {
    constructor(driver, queryConfig) {
        this.cursor = "";
        this.driver = driver;
        this.queryConfig = queryConfig;
    }
    /** Queries events since last call to queryEvents */
    async queryEvents(labels, txFilter) {
        if (!this.cursor) {
            this.cursor = await this.fetchCurrentChangeId();
        }
        const cursorLiteral = new cypher_builder_1.default.Literal(this.cursor);
        const selectors = this.createQuerySelectors(labels);
        if (txFilter) {
            selectors.map((selector) => {
                selector.set("txMetadata", txFilter);
            });
        }
        const queryProcedure = cypher_builder_1.default.db.cdc.query(cursorLiteral, selectors);
        try {
            const events = await this.runProcedure(queryProcedure);
            this.updateChangeIdWithLastEvent(events);
            return events;
        }
        catch (err) {
            if (err instanceof neo4j_driver_1.Neo4jError) {
                // Cursor is stale, needs to be reset
                // Events between this error and the next poll will be lost
                if ((0, error_has_gql_status_1.errorHasGQLStatus)(err, "52N29") ||
                    err.code === "Neo.ClientError.ChangeDataCapture.InvalidIdentifier") {
                    console.warn(err);
                    this.cursor = ""; // Resets cursor
                    return [];
                }
            }
            throw err;
        }
    }
    async updateCursor() {
        this.cursor = await this.fetchCurrentChangeId();
    }
    async fetchCurrentChangeId() {
        const currentProcedure = cypher_builder_1.default.db.cdc.current();
        const result = await this.runProcedure(currentProcedure);
        if (result[0] && result[0].id) {
            return result[0].id;
        }
        else {
            throw new Error("id not available on cdc.current");
        }
    }
    updateChangeIdWithLastEvent(events) {
        const lastEvent = events[events.length - 1];
        if (lastEvent) {
            this.cursor = lastEvent.id;
        }
    }
    createQuerySelectors(labels) {
        if (labels) {
            return labels.map((l) => new cypher_builder_1.default.Map({
                select: new cypher_builder_1.default.Literal("n"),
                labels: new cypher_builder_1.default.Literal([l]),
            }));
        }
        else {
            // Filters nodes
            return [
                new cypher_builder_1.default.Map({
                    select: new cypher_builder_1.default.Literal("n"),
                }),
            ];
        }
    }
    async runProcedure(procedure) {
        const { cypher, params } = procedure.build();
        const result = await this.driver.executeQuery(cypher, params, this.queryConfig);
        return result.records.map((record) => {
            return record.toObject();
        });
    }
}
exports.CDCApi = CDCApi;
